### ISSUE TITLE
Moving averages, line smoothing, etc.

### ISSUE THREAD

**Issue by mbostock** [d3/d3-shape#43]
I’m not 100% sure this belongs as a curve (or as a data transformation), but it seems like a reasonable place to put it. In theory if you used a smooth curve, the input data array to the line or area generator wouldn’t even need to be sorted by *x*-value: the smoother could buffer all the points and then compute the smooth curve on lineEnd.

**Comment by mbostock** [comment]
A kernel regression is another possibility, as in this [kernel density estimation](http://bl.ocks.org/mbostock/4341954) example. That implementation is O(n^2), but I suspect you could make it O(n lg n), or even O(n) if you assume that data is already sorted on *x*.

**Comment by mbostock** [comment]
[LOESS](https://en.wikipedia.org/wiki/Local_regression) is another option.

**Comment by mbostock** [comment]
And here’s a simple test case:

[![screen shot 2015-12-01 at 1 17 59 pm](https://cloud.githubusercontent.com/assets/230541/11514564/fa34353a-982d-11e5-8a4b-c1fb3efbaa2f.png)](http://bl.ocks.org/mbostock/d8bcc4b130df420d6c40)

With a smooth curve in R:

[![cvgpl7jueaamixy](https://cloud.githubusercontent.com/assets/230541/11514587/1a15c4ae-982e-11e5-8274-8b31d3e73a7e.png)](https://twitter.com/enf/status/671475255202967553)

**Comment by curran** [comment]
Yes! When I first saw this visualization I thought of trying to implement a gaussian KDE for use with D3.

**Comment by mbostock** [comment]
I couldn’t resist taking a crack at it. Here’s the slow O(n^2) implementation of kernel smoothing:

[![screen shot 2015-12-01 at 3 34 12 pm](https://cloud.githubusercontent.com/assets/230541/11517825/008ec70c-9841-11e5-8652-f0d3782d3be5.png)](http://bl.ocks.org/mbostock/963c4d246e96963856bf)

Things that would be nice:

- [ ] Make it faster than O(n^2).
- [ ] Support kernels besides Epanechnikov.
- [ ] Choose the bandwidth automatically?
- [ ] Choose the precision (number of points in generated curve) automatically?
- [ ] Don’t depend on d3_scale (generating irregular ticks would be fine).

**Comment by mbostock** [comment]
I’m going back to working on API documentation if anyone else wants to make a better implementation…

**Comment by curran** [comment]
Doing a little research, here's a [list of common kernel functions](https://en.wikipedia.org/wiki/Kernel_(statistics)#Kernel_functions_in_common_use)

* [ ] Uniform
* [ ] Triangular
* [ ] Epanechnikov  
* [ ] Quartic 
* [ ] Triweight   
* [ ] Tricube
* [ ] Gaussian
* [ ] Cosine
* [ ] Logistic
* [ ] Silverman 

Also, here's a [formula for bandwidth estimation using "Silverman's rule of thumb"](https://en.wikipedia.org/wiki/Kernel_density_estimation#Practical_estimation_of_the_bandwidth).

**Comment by mbostock** [comment]
See also [science.js](https://github.com/jasondavies/science.js/tree/master/src/stats).

**Comment by ericfischer** [comment]
My [fit above](https://twitter.com/enf/status/671475255202967553), by the way, is the sum of two gaussian PDFs, plus the sum of the same gaussians for x-24 and x+24, all raised to a power, for no good reason other than that it looked about right. I'd love to know a better way than eyeballing it and then applying general curve fitting to guess the components of things that look like the sum of gaussians.

**Comment by mbostock** [comment]
See also ggplot2’s [stat_smooth](http://docs.ggplot2.org/current/stat_smooth.html), which uses LOESS and [GAM](https://en.wikipedia.org/wiki/Generalized_additive_model) [(generalized additive models)](http://www.inside-r.org/r-doc/mgcv/gam).

ggplot2 also has a [geom_smooth](http://docs.ggplot2.org/current/geom_smooth.html) which can smooth using a ribbon, showing an upper and lower bound on the smoothed model. That might be a nice curve type for an area shape.

**Comment by curran** [comment]
Here's the [Loess + D3 example from science.js](https://github.com/jasondavies/science.js/tree/master/examples/loess) by @jasondavies . I made a bl.ock to study it: 

[![loess](https://cloud.githubusercontent.com/assets/68416/11519074/d27ea61a-984b-11e5-873e-1c1565a0618b.png)](http://bl.ocks.org/curran/890f8475dfe06710ee62)

It looks like the API expects `loess(xValuesArray, yValuesArray)` and returns an array of smoothed Y values.

**Comment by mbostock** [comment]
Since the kernel operates in pixel space, I bet it would be reasonable to just have stupid defaults for the bandwidth (±20px) and precision (control points every 10px). Though certainly it would be nice if Simpson’s rule of thumb were easy to implement if you wanted a better default value, albeit at the cost of a global computation.

**Comment by curran** [comment]
Here's an example that adds a LOESS curve to your tweet example using science.js:
[![loess](https://cloud.githubusercontent.com/assets/68416/11542408/0db858fc-98ec-11e5-9aee-26ab4f09dc68.png)](http://bl.ocks.org/curran/320c765d77cbb5b37e08)

I made this to try to understand the science.js LOESS implementation. Here are a few things I learned:

 * It was necessary to sort the data by time. I thought it would be sorted by time already (based on the order from the CSV file), but somehow it doesn't work without first sorting by time.
 * The `loess` function accepts arrays of values in data space, and surprisingly it works with dates.
 * The `loess` function returns an arrays of smoothed Y values that is the same length as the original arrays. I would have hoped that the length of the returned array is configurable, but it looks like in this implementation it is not.

**Comment by mbostock** [comment]
I think it might be a good idea to do smoothing as a pre-processing step in data space rather than doing it internally as a curve:

* Data space seems a more natural fit when specifying kernel bandwidth.
* You might want to compute a smooth fit once while interactively redrawing (e.g., pan & zoom).
* You might want to access points in the computed smooth fit for display (e.g., hover tooltip).

I’m imagining something that has a *x* and *y* accessors and other configuration (e.g., kernel and bandwidth) and then computes the smooth curve, generating an array of [[x1, y1], [x2, y2], …] points in data space. Those can then be passed to a line generator, likely with an *x* and *y* accessor that apply scales to transform to screen space.

(I’m slightly tempted to have it return [[x1, x2, x3, …], [y1, y2, y3, …]] because that’s a more more efficient representation, but it would make implementing the accessors a little bit more awkward, so it feels like premature optimization.)

**Comment by tomgp** [comment]
+1 on the idea that this should be a data space operation.

**Comment by mbostock** [comment]
Probably should create a d3-smooth module for smoothing functions.

**Comment by muonsw** [comment]
Hope you don't mind me commenting but I found my way to this after looking at the issues on science.js.

I'm sure things have moved on since December, but I implemented a kernel smoothing function for use with D3 a while back. It cuts off the calculation for each point once the calculated weight goes below a defined threshold (with relation to the weight at the point that is being smoothed).

- It looks like it scales linearly but there is a trade-off in accuracy. So for the twitter data in question it’s about an order of magnitude quicker (depending on the bandwidth) but gives a line that can be up to 0.03% different.
- It assumes that data is sorted on the x-coordinate and hence that the calculation can be cut off.
- It assumes that the kernel function will always decrease beyond the cut-off point.

I'm fairly new to JavaScript so I'm sure that it’s not the best implementation. Anyway, I've created a [gist](https://bl.ocks.org/muonsw/3eb574d4b247fb4a484b) if it’s of any interest (or use).

![gist](https://cloud.githubusercontent.com/assets/16984337/13630194/f8ff3eee-e5d5-11e5-853d-f8770b6a292d.png)

**Comment by nealstewart** [comment]
@mbostock Any thoughts in this area? I need some line smoothing for a time series of continuous data in my application, and would love to contribute something.

**Comment by mbostock** [comment]
Here’s a simple moving average implementation:

https://observablehq.com/@d3/moving-average

**Comment by mbostock** [comment]
This feels more appropriate to do in data space rather than geometrically, so I don’t think d3-shape is the right place for this. Still interested in seeing progress here, though.
